[Messenger] Add `WorkerMetadata` to `Worker` class.

okwinza opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fixes #37736
| License       | MIT
| Doc PR        | -

At the moment, there is no clean way to access the values of `transportNames` or recently introduced `queueNames` that the worker was configured with, although such data might be quite useful for logging/monitoring or other tasks.

This PR attempts to fix that by adding a new and extensible way to provide additional information about a particular `Worker` object.

So far, the following PRs could benefit from this change:

- #43133 
- #42723

**Use case example:**
----

- As I developer
- When a message was consumed from transport with name `async`.
- And the worker state is `idle`.
- Then I want to reset services.

**Before this PR**, the only solution not relying on using Reflection API would look like this:
```php
    private $servicesResetter;
    private $receiversName;
    private $actualReceiverName = null;

    public function __construct(ServicesResetter $servicesResetter, array $receiversName)
    {
        $this->servicesResetter = $servicesResetter;
        $this->receiversName = $receiversName;
    }

    public function saveReceiverName(AbstractWorkerMessageEvent $event): void
    {
        $this->actualReceiverName = $event->getReceiverName();
    }

    public function resetServices(WorkerRunningEvent $event): void
    {
        if (!$event->isWorkerIdle() && \in_array($this->actualReceiverName, $this->receiversName, true)) {
            $this->servicesResetter->reset();
        }

        $this->actualReceiverName = null;
    }

    public static function getSubscribedEvents(): array
    {
        return [
            WorkerMessageHandledEvent::class => ['saveReceiverName'],
            WorkerMessageFailedEvent::class => ['saveReceiverName'],
            WorkerRunningEvent::class => ['resetServices'],
        ];
    }
```
**With this PR**, one could simply use this to retrieve the transport name.
```php
$event->getWorker()->getWorkerMetadata()->getTransportName() === $this->transportName;
```
So the whole solution would look like this:
```php
    private $servicesResetter;
    private $receiversName;

    public function __construct(ServicesResetter $servicesResetter, array $receiversName)
    {
        $this->servicesResetter = $servicesResetter;
        $this->receiversName = $receiversName;
    }

    public function resetServices(WorkerRunningEvent $event): void
    {
        $actualTransportName = $event->getWorker()->getWorkerMetadata()->getTransportName();
        if (!$event->isWorkerIdle() || !in_array($actualTransportName, $this->receiversName, true)) {
            return;
        }

        $this->servicesResetter->reset();
    }

    public static function getSubscribedEvents(): array
    {
        return [
            WorkerRunningEvent::class => ['resetServices'],
        ];
    }
```

alexander-schranz commented:
@carsonbot find another a reviewer please.

okwinza commented:
Friendly ping @Nyholm. 
Could you see any value in such a feature? :)

Guikingone commented:
Hi 👋🏻 

I may ask a strange question but what's the benefits compared to using a `getWorkerMetadata(): array`? 

As an event already receive the `Worker`, any benefits from using a dedicated class? 🤔

okwinza commented:
> Hi 👋🏻
> 
> I may ask a strange question but what's the benefits compared to using a `getWorkerMetadata(): array`?
> 
> As an event already receive the `Worker`, any benefits from using a dedicated class? 🤔

Hi, indeed, the whole PR might be replaced with one or a couple getters but:
1) IMO bloating `Worker` class with getters sounds more like a hack and doesn't seem like a good idea overall, it's better to keep the concerns separated here.
2) I considered going with `getWorkerMetadata(): array` but the DX looked much worse to me since it wasn't clear what's being returned without first looking through the code/docs, etc. Not to mention that IDE autocomplete wont work and you won't be able to enforce return types on the language level that way. I know we have psalm for this, but still.

chalasr commented:
I like the idea of using an object but I think it's fine to make it accessible from the Worker instead of passing it to all worker-aware events.

okwinza commented:
@chalasr I've refactored the PR according to your suggestions. Should be good to go once the CI is green.

okwinza commented on src/Symfony/Component/Messenger/WorkerMetadata.php:
```diff
@@ -0,0 +1,49 @@
+        $this->metadata = $metadata;
+    }
+
+    public function set(string $key, $value): void
```
Since populating the object is purely an internal operation, I don't think it makes sense to pollute the class with a bunch of individual setters for each property here. Hence the generic `set` method.

okwinza commented:
Ping @chalasr @derrabus.
Is there anything I could do to make the review easier for you guys? :)

lyrixx commented on src/Symfony/Component/Messenger/WorkerMetadata.php:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Returns the queue names the worker consumes from, if "--queues" option was used.
+     * Returns false otherwise.
+     *
+     * @return array|false
+     */
+    public function getQueueNames()
+    {
+        return $this->metadata['queueNames'] ?? false;
```
```suggestion
     */
    public function getQueueNames(): ?array
    {
        return $this->metadata['queueNames'] ?? null;
```

this version feels more natural to me

okwinza replied:
I totally agree with you here. The reason why I went with `false` is to keep things consistent with this: https://github.com/symfony/symfony/blob/9260aa6607992503b8063a690692f9ec545b1a07/src/Symfony/Component/Messenger/Worker.php#L73

But I guess we could just change the both cases to `?? null` instead. :)

lyrixx reviewed (commented):
It looks cool.

But one question:

from where you gonna call `$worker->getWorkerMetadata()` ?

okwinza commented:
> It looks cool.
> 
> But one question:
> 
> from where you gonna call `$worker->getWorkerMetadata()` ?

Much thanks for the review @lyrixx, appreciate it!
The primary use case is to call it from any worker event subscriber,  mostly via `WorkerRunningEvent::getWorker()` or any other event that contains a reference to a Worker object.

okwinza committed:
[Messenger] Add WorkerMetadata to Worker class

Add tests for WorkerMetadata class


Fix WorkerMetadataTest header


Access WorkerMetadata directly via getter and make it mutable.


Fix CS


Refactor WorkerMetadata::set() method to accept array


Use null as default value for queueNames instead of false


Add missing return type


Make fabbot happy

okwinza commented:
Rebased, squashed & updated the PR description. Should be good to go now :partying_face: .

lyrixx commented:
Thanks for your work on this new feature!
